src/mgr/DaemonServer.cc

```diff
@@ -390,14 +390,11 @@ bool DaemonServer::handle_report(MMgrReport *m)
     dout(20) << "updating existing DaemonState for " << key << dendl;
     daemon = daemon_state.get(key);
   } else {
-    dout(4) << "constructing new DaemonState for " << key << dendl;
-    daemon = std::make_shared<DaemonState>(daemon_state.types);
-    // FIXME: crap, we don't know the hostname at this stage.
-    daemon->key = key;
-    daemon_state.insert(daemon);
-    // FIXME: we should avoid this case by rejecting MMgrReport from
-    // daemons without sessions, and ensuring that session open
-    // always contains metadata.
+    // we don't know the hostname at this stage, reject MMgrReport here.
+    dout(4) << "rejecting report from non-daemon client " << m->daemon_name
+	    << dendl;
+    m->put();
```

Comment:
we need to kill the session at this point

could we do it as following:

```diff
diff --git a/src/mgr/DaemonServer.cc b/src/mgr/DaemonServer.cc
index a867754..8502f72 100644
--- a/src/mgr/DaemonServer.cc
+++ b/src/mgr/DaemonServer.cc
@@ -391,10 +391,17 @@ bool DaemonServer::handle_report(MMgrReport *m)
     daemon = daemon_state.get(key);
   } else {
     // we don't know the hostname at this stage, reject MMgrReport here.
-    dout(4) << "rejecting report from non-daemon client " << m->daemon_name
+    dout(1) << "rejecting report from " << k ", since we do not have its metadata now."
            << dendl;
-    m->put();
-    return true;
+    // kill session
+    MgrSessionRef session(static_cast<MgrSession*>(m->get_connection()->get_priv()));
+    if (!session) {
+      return false;
+    }
+    m->get_connection()->mark_down();
+    session->put();
+
+    return false;
   }

```